bitsail-connectors/connector-ftp/src/main/java/com/bytedance/bitsail/connector/ftp/source/reader/FtpSourceReader.java

```diff
@@ -0,0 +1,190 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.bytedance.bitsail.connector.ftp.source.reader;
+
+import com.bytedance.bitsail.base.connector.reader.v1.SourcePipeline;
+import com.bytedance.bitsail.base.connector.reader.v1.SourceReader;
+import com.bytedance.bitsail.base.format.DeserializationSchema;
+import com.bytedance.bitsail.common.BitSailException;
+import com.bytedance.bitsail.common.configuration.BitSailConfiguration;
+import com.bytedance.bitsail.common.row.Row;
+import com.bytedance.bitsail.connector.ftp.core.client.FtpHandlerFactory;
+import com.bytedance.bitsail.connector.ftp.core.client.IFtpHandler;
+import com.bytedance.bitsail.connector.ftp.core.config.FtpConfig;
+import com.bytedance.bitsail.connector.ftp.error.FtpErrorCode;
+import com.bytedance.bitsail.connector.ftp.source.split.FtpSourceSplit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+
+public class FtpSourceReader implements SourceReader<Row, FtpSourceSplit> {
+  private static final Logger LOG = LoggerFactory.getLogger(FtpSourceReader.class);
+
+  private final int subTaskId;
+
+  private final IFtpHandler ftpHandler;
+  private transient BufferedReader br;
+  private int totalSplitNum = 0;
+  private boolean hasNoMoreSplits = false;
+
+  private final Deque<FtpSourceSplit> splits;
+  private final transient DeserializationSchema<byte[], Row> deserializationSchema;
+
+  private FtpSourceSplit currentSplit;
+  private long currentReadCount;
+  private transient String line;
+
+  private FtpConfig ftpConfig;
+
+  private boolean skipFirstLine = false;
+  private String successFilePath;
+  private final transient Context context;
+
+  public FtpSourceReader(BitSailConfiguration jobConf, Context context, int subTaskId) {
+    this.subTaskId = subTaskId;
+    this.ftpConfig = new FtpConfig(jobConf);
+    this.ftpHandler = FtpHandlerFactory.createFtpHandler(ftpConfig);
+    this.splits = new LinkedList<>();
+    this.context = context;
+    this.deserializationSchema = DeserializationSchemaFactory.createDeserializationSchema(jobConf, ftpConfig, context);
+    LOG.info("FtpReader is initialized.");
+
+  }
+
+  @Override
+  public void start() {
+    this.ftpHandler.loginFtpServer();
+    checkPathsExist();
```

Comment:
I think we could move the folder check to the coordinator, and we could check if the path of the split exists in the reader.